_moderngl.py

```diff
@@ -366,5 +366,168 @@ def make_varying(name, number, array_length, dimension):
+
+    pointer_variable = {}  # id : pointer_type_id 
+    pointer_allowed_types = {}  # pointer_type_id : type_id
+    allowed_types = {}  # type_id : [py_type, type_id]
+
+    extracted_types = {}  # id : py_type
+    extracted_location = {}  # id : location
+    while idx < len(spv):
+        args, opcode = spv[idx] >> 16, spv[idx] & 0xffff
+        if opcode == 5: # OpName
+            # We can extract the name of some ids
+            extracted_names[spv[idx + 1]] = raw_spv[idx*4 + 2*4 : idx*4 + args*4].decode()
+        
+        if opcode == 59: # OpVariable
+            # We can extract if it is a vertex shader input or not
+            extracted_storage_class_id[spv[idx + 2]] = spv[idx + 3]
+            
+            # Mapping pointer_type_id to id of user variables
+            pointer_variable[spv[idx + 2]] = spv[idx + 1]
+        
+        if opcode == 71: # OpDecorate
+            # We can extract the location here
+            extracted_location[spv[idx + 1]] = spv[idx + 3]
+        
+        if opcode == 32:  # OpTypePointer
+            # Retrieving used (allowed) type_ids for pointer_type_ids
+            pointer_allowed_types[spv[idx + 1]] = spv[idx + 3]
+            # --------------------^pointer_id^^----^^^type_id^^==
+            #print(spv[idx + 1], spv[idx + 2], spv[idx + 3])
+        
+        if opcode == 19:  # OpTypeVoid
+            allowed_types[spv[idx + 1]] = ['void', -1]
+        if opcode == 20:  # OpTypeBool
+            allowed_types[spv[idx + 1]] = ['bool', -1]
+         
+        if opcode == 21:  # OpTypeInt
+            addware_unsigned = ''
+            if spv[idx + 3] == 1:
+                 addware_unsigned = 'u'
+            allowed_types[spv[idx + 1]] = [f'{addware_unsigned}i{spv[idx + 2]//8}', -1]
+            
+        if opcode == 22:  # OpTypeFloat
+            allowed_types[spv[idx + 1]] = [f'f{spv[idx + 2]//8}', -1]
+            
+        if opcode == 23:  # OpTypeVector
+            allowed_types[spv[idx + 1]] = [f'vec{spv[idx + 3]}', spv[idx + 2]]
+            
+        if opcode == 24:  # OpTypeMatrix
+            allowed_types[spv[idx + 1]] = [f'mat{spv[idx + 3]}', spv[idx + 2]]
+        
+        idx += args
+    ##################
+
+    # == Assembly types == #
+    # Some types of variables have pointers to other types of variables used in them.
+    # So we use endless research to identify complete types of variables.
+    def assembly(type_id):
+        typ, thrw_typ = allowed_types[type_id]
+        if thrw_typ != -1:
+            add_typ = assembly(thrw_typ)
+            allowed_types[type_id][0] = add_typ+' '+typ
+            allowed_types[type_id][1] = -1
+        return typ
+
+    for type_id, config in allowed_types.items():
+        assembly(type_id)
+
+    for ids, pointer_type_id in pointer_variable.items():
+        try:
+            extracted_types[ids] = allowed_types[pointer_allowed_types[pointer_type_id]][0]
+        except KeyError:
```

Comment:
when do you get a key error?

Reply:
There have been cases when built-in variables appeared when searching for names. As I recently said gl_InstanceID and the like. 
Well, some of them did not have their typing, and, accordingly, they were not in the allowed (that is, mentioned in SPIR-V) types, so I put such an exception.